amy/workshops/forms.py

```diff
@@ -564,22 +564,24 @@ def clean_curricula(self):
         """Validate tags when some curricula are selected."""
         curricula = self.cleaned_data["curricula"]
         tags = self.cleaned_data["tags"]
-
         try:
-            expected_tags = []
+            expected_tags = set()
             for c in curricula:
                 if c.active and c.carpentry:
-                    expected_tags.append(c.carpentry)
+                    expected_tags.add(c.carpentry)
                 elif c.active and c.mix_match:
-                    expected_tags.append("Circuits")
+                    expected_tags.add("Circuits")
         except (ValueError, TypeError):
-            expected_tags = []
+            expected_tags = set()
 
+        missing_tags = set()
         for tag in expected_tags:
             if not tags.filter(name=tag):
-                raise forms.ValidationError(
-                    "You must add tags corresponding to these curricula."
-                )
+                missing_tags.add(tag)
+        if missing_tags:
+            raise forms.ValidationError(
+                f"You must add tags corresponding to these curricula. Missing tags: {', '.join(missing_tags)}"
+            )
```

Comment:
So this method `clean_curricula` validates tags based on selected curricula. I think it should be `clean_tags`, so that at least the error shows up in proper place.

Also it could be beneficial for user experience if the fields in the form switched places.

Reply:
I agree - though updating this was less straightforward than I thought. In short, `curricula` wasn't in `cleaned_data` at the time `clean_tags` was run during validation, so I've changed this method to be called during `clean()` - following these Django docs https://docs.djangoproject.com/en/4.1/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other. Can you do a quick re-review?

Reply:
@elichad Looks okay except for the way validation error is handled.

We have this pattern:

```python
errors = {}

if validation_failed_for_a_field:
    errors["field"] = ValidationError("Text")

# ...

if errors:
    raise ValidationError(errors)
```

For example: `CommunityRoleForm.clean`, `amy/communityroles/forms.py`.

Reply:
Ah, thanks for the reference, done